theories/datatypes/MSet.ec

```diff
@@ -0,0 +1,689 @@
+proof. by rewrite /mset1 oflist_mult /= /pred1. qed.
+
+op [opaque] add ['a] (s1 s2 : 'a mset) = oflist (elems s1 ++ elems s2).
+
+abbrev (+) ['a] = add<:'a>.
+
+lemma add_mult (s1 s2: 'a mset) e: (s1 + s2).[e] = s1.[e] + s2.[e].
+proof. by rewrite /add oflist_mult count_cat /mult. qed.
+
+op [opaque] union ['a] (s1 s2 : 'a mset) = offmap (merge 
+  (fun a m n, if is_some m \/ is_some n then Some (max (odflt 0 m) (odflt 0 n)) else None) 
+  (tofmap s1) (tofmap s2)).
+
+abbrev (`|`) ['a] = union<:'a>.
+
+lemma union_mult (s1 s2: 'a mset) e: (s1 `|` s2).[e] = max s1.[e] s2.[e].
+proof. rewrite /union offmap_mult mergeE //= !tofmap_get; smt(mult_ge0 mult0). qed.
+
+op [opaque] intersection ['a] (s1 s2 : 'a mset) = offmap (merge 
+  (fun a m n, if is_some m /\ is_some n then Some (min (odflt 0 m) (odflt 0 n)) else None) 
+  (tofmap s1) (tofmap s2)).
+
+abbrev (`&`) ['a] = intersection<:'a>.
+
+lemma intersection_mult (s1 s2: 'a mset) e: (s1 `&` s2).[e] = min s1.[e] s2.[e].
+proof. rewrite /intersection offmap_mult mergeE //= !tofmap_get; smt(mult_ge0 mult0). qed.
+
+op [opaque] diff ['a] (s1 s2 : 'a mset) = offmap (merge 
+  (fun a m n, if is_some m then Some (odflt 0 m - odflt 0 n) else None) 
+  (tofmap s1) (tofmap s2)).
+
+abbrev (`\`) ['a] = diff<:'a>.
+
+lemma diff_mult (s1 s2: 'a mset) e: (s1 `\` s2).[e] = max 0 (s1.[e] - s2.[e]).
+proof. rewrite /diff offmap_mult mergeE //= !tofmap_get; smt(mult_ge0 mult0). qed.
+
+op rem ['a] (s : 'a mset) x = s `\` mset1 x.
+
+lemma rem_mult s (x y: 'a): (rem s x).[y] = max 0 (s.[y] - b2i (x = y)).
+proof. by rewrite diff_mult mset1_mult. qed.
+
+op insert ['a] x (s : 'a mset) = s + mset1 x.
+
+lemma insert_mult s (x y: 'a): (insert x s).[y] = s.[y] + b2i (x = y).
+proof. by rewrite add_mult mset1_mult. qed.
+
+(* -------------------------------------------------------------------- *)
+
+lemma elems_mset0 ['a]: elems mset0 = [<:'a>].
+proof. 
+rewrite /mset0; apply/perm_eq_small/perm_eq_sym => //=.
+rewrite -{1}(undup_id []) // oflistK //.
+qed.
+
+lemma elems_eq_mset0 ['a] (A : 'a mset): elems A = [<:'a>] => A = mset0.
+proof.
+move=> elems_nil; apply/mset_eqP=> x. 
+by rewrite mset0_mult -mult0 /mem elems_nil.
+qed.
+
+lemma elems_mset1 (x : 'a) : elems (mset1 x) = [x].
+proof.
+rewrite /mset1; apply/perm_eq_small/perm_eq_sym=> //=.
+by rewrite -{1}(undup_id [x]) ?oflistK.
+qed.
+
+(* -------------------------------------------------------------------- *)
+
+hint rewrite multE : mset0_mult mset1_mult add_mult union_mult  
+  intersection_mult diff_mult.
+
+lemma oflist_cons ['a] (x : 'a) s : oflist (x::s) = insert x (oflist s).
+proof. 
+apply mset_eqP => z.
+rewrite oflist_mult /= !multE /mult {1}/pred1 addzC.
+congr; apply perm_eqP_pred1.
+exact oflistK.
+qed.
+
+(* -------------------------------------------------------------------- *)
+abbrev disjoint (xs ys : 'a mset) = xs `&` ys = mset0.
+
+lemma disjointP (xs ys : 'a mset):
+  disjoint xs ys <=> forall (x : 'a), x \in xs => ! x \in ys.
```

Comment:
The right hand side could be an equivalence.

I'd recommend showing that `disjoint` is commutative, keeping `disjointP` as it is, and additionally proving a version that gives out a membership equivalence characterisation from `disjoint`.

Reply:
There isn't an equivalence. Elements in one set can't be in the other, but there can be elements that are in neither.

I'll add the commutativity lemma though, which is immediate from `msetIC`.